jetbrains-core/src/software/aws/toolkits/jetbrains/services/s3/editor/S3TreeNode.kt

```diff
@@ -3,29 +3,37 @@
 
 package software.aws.toolkits.jetbrains.services.s3.editor
 
+import com.intellij.icons.AllIcons
 import com.intellij.openapi.fileTypes.FileTypeRegistry
-import com.intellij.openapi.fileTypes.UnknownFileType
 import com.intellij.openapi.util.io.FileUtilRt
 import com.intellij.ui.treeStructure.SimpleNode
 import kotlinx.coroutines.runBlocking
-import software.amazon.awssdk.services.s3.model.ListObjectVersionsResponse
 import software.aws.toolkits.jetbrains.services.s3.NOT_VERSIONED_VERSION_ID
 import software.aws.toolkits.resources.message
 import java.time.Instant
 
-sealed class S3TreeNode(val bucketName: String, val parent: S3LazyLoadParentNode?, val key: String) : SimpleNode() {
-    open val isDirectory = false
+sealed class S3TreeNode(val bucket: S3VirtualBucket, val parent: S3LazyLoadParentNode?, val key: String) : SimpleNode() {
     override fun getChildren(): Array<S3TreeNode> = arrayOf()
-    override fun getName(): String = key.substringAfterLast('/')
-}
 
-fun S3TreeNode.getDirectoryKey() = if (isDirectory) {
-    key
-} else {
-    parent?.key ?: throw IllegalStateException("$key claimed it was not a directory but has no parent!")
+    @Deprecated("Do not use, exists due to SimpleNode. Use more use case specific methods", ReplaceWith("displayName()"))
+    final override fun getName(): String = displayName()
+
+    /**
+     * String representation of this node in UIs
+     */
+    open fun displayName(): String = key.trimEnd('/').substringAfterLast('/')
+
+    /**
+     * Directory path to this node
+     */
+    open fun directoryPath() = parent?.key ?: throw IllegalStateException("$key has no parent!")
+
+    override fun toString(): String = "${this::class.simpleName}(key='$key')"
+
+    override fun getEqualityObjects(): Array<Any> = arrayOf(bucket, key)
```

Comment:
what system uses this? is it okay if a subclass forgets to override with something more specific?

Reply:
SimpleNode uses it for its equals hashcode...

Without it we cant write assertThat statements without custom matchers.